SEP-31: integrate SEP-38 (#932)
stellar/stellar-protocol: ecosystem/sep-0031.md

- `application/json`

## Asset Conversions and Liquidity Strategies

> This protocol can be used if the Sending and Receiving Client wish to send and receive the same asset, but this is typically not the case. It is more common for the Sending Client to provide the Sending Anchor with a particular asset (the source asset) and have the Receiving Client receive a different asset (the destination asset).

Reviewer: 💡 This paragraph is a little difficult to digest because it leads with the least important information and finishes with the most important information. The most important information is what the protocol is intended to be used for. Suggest reordering:
```suggestion
This protocol is designed for the common case where the Sending Client provides the Sending Anchor with a particular asset (the source asset) and the Receiving Client receives a different asset (the destination asset). It is possible to use this protocol when the Sending and Receiving clients wish to send and receive the same asset, but that is not the common use case it has been designed.
```

> In this case, the source asset must convert into the receiving asset at some point prior to delivering funds to the Receiving Client. This conversion must be done by either the Sending Anchor, the Stellar Network, or the Receiving Anchor. The subsections below outline the stratgies used for each method.

Reviewer: >In this case,

💡  It's ambiguous at the start of the sentence what "this case" is because the prior paragraph discusses two cases. The reader has to look ahead towards to rest of the paragraph to figure it out. I think this would make the paragraph clearer:

```suggestion
When the source asset and destination assets are different, the source asset must convert into the receiving asset at some point prior to delivering funds to the Receiving Client. This conversion must be done by either the Sending Anchor, the Stellar Network, or the Receiving Anchor. The subsections below outline the strategies used for each method.
```

🔠 I also fixed a typo in this paragraph.

### Sending Anchor Asset Conversion

The Sending Anchor can collect the source asset from the Sending Client and send the destination asset to the Receiving Anchor. 

This approach requires the Sending Anchor to provide an exchange rate, or quote, to the Sending Client. It also requies the Sending Anchor to hold a balance of the destination asset on Stellar. The Sending Anchor would use Payment operations to send the quoted amount of the destination asset to the Receiving Anchor.

### Stellar Network Asset Conversion

> Instead of holding a balance of the destination asset on Stellar, the Sending Anchor can instead us Path Payments to convert the source asset to the destination asset on the Stellar Network. 

Reviewer: 🔠  There's a double instead in this sentence, and a typo.
```suggestion
Instead of holding a balance of the destination asset on Stellar, the Sending Anchor can use Path Payments to convert the source asset to the destination asset on the Stellar Network. 
```

> This approach requires a sufficiently liquid market to exist between the source and destination assets on the Stellar Decentralized Exchange (SDEX). Note that the exchange rate between the two assets is dynamic due to market fluctuations on the SDEX.

Reviewer: >Note that the exchange rate between the two assets is dynamic due to market fluctuations on the SDEX.

❓ This is mentioned here as if it is unique to the SDEX, but in all cases the rate between two assets will be fluid. What's the intention with calling it out for only the SDEX?

Author: Fair point. I can remove it 👍

### Receiving Anchor Asset Conversion

Finally, the Receiving Anchor can receive payments of the source asset on Stellar and deliver payments of the destination asset to the Receiving Client. 

> This requires the Receiving Anchor to implement the [SEP-38 Anchor RFQ API](sep-0038.md), which enables the Receiving Anchor to provide quotes to the Sending Anchor. These quotes can be [indicative](https://www.investopedia.com/terms/i/indicativequote.asp) or [firm](https://www.investopedia.com/terms/f/firmquote.asp). If a firm quote is used, the Stellar transaction _must_ be submitted to the Stellar network prior to the quote's expiration. More specifically, the `created_at` timestamp returned in the [Horizon GET /transactions/:id](https://developers.stellar.org/api/resources/transactions/) request must be earlier than the quote's expiration.

Reviewer: 💭 There are some subtleties regarding submitting transactions, the created at timestamp, and how long they are valid and can be resubmitted that we might need to consider in this scenario.

The transaction not only needs to be submitted before the quote's expiration, but it needs to be included in a closed ledger. The created at timestamp on transactions is the nomination time of the ledger (the time the ledger closed) that the transaction was included in. It will be possible for a transaction to be queued, and so it could be submitted prior to the expiry time but not included in the next ledger but rather a future ledger until after the close time. This may be confusing to implementers. This is an edge case, but we should probably still document what the two parties should do if this happens. Implementers could also reduce the chance of this happening by setting the time bounds on the transaction to be the expiry time or earlier.

Author: > Implementers could also reduce the chance of this happening by setting the time bounds on the transaction to be the expiry time or earlier.

If they do this, there is actually _no_ chance a transaction would be submitted earlier and included later than the expiration right? That detail may be important enough to explicitly mention here.

Reviewer: I think it's a little fuzzy, I'd verify it with the core team, but essentially yes.

> 

Reviewer 2: This is really great. We should have @phil-stellar and @lisaclaire828 take a look as well. I would consider the following renaming of strategies to reflect *when* the asset conversion happens: 

`Sending Anchor Asset Conversion` -> `Pre-payment asset conversion`

`Stellar Network Asset Conversion`-> `Real time asset conversion` 

`Receiving Anchor Asset Conversion` -> `Post-payment asset conversion`

Reviewer: I think those names somewhat obfuscate who is responsible for asset conversion, and if I read them without context of what's in each section I don't really know what they mean.

Author: I agree with Leigh, I think the _who_ is more important than the _when_

### Sending Anchor Asset Conversion

The Sending Anchor can collect the source asset from the Sending Client and send the destination asset to the Receiving Anchor. 

This approach requires the Sending Anchor to provide an exchange rate, or quote, to the Sending Client. It also requires the Sending Anchor to hold a balance of the destination asset on Stellar. The Sending Anchor would use Payment operations to send the quoted amount of the destination asset to the Receiving Anchor.

### Stellar Network Asset Conversion

Instead of holding a balance of the destination asset on Stellar, the Sending Anchor can use Path Payments to convert the source asset to the destination asset on the Stellar Network. 

This approach requires a sufficiently liquid market to exist between the source and destination assets on the Stellar Decentralized Exchange (SDEX).

> ### Receiving Anchor Asset Conversion

Reviewer 3: Q: is this type of conversion a new addition in this version of the protocol? How many Anchors (ballpark estimation) do you think are using SEP-31 already?

Author: Yes, prior to the development of SEP-38 and these changes to SEP-31 it wasn't possible for the receiving anchor to convert assets in a standardized way. Currently, every anchor that implements SEP-31 is interested in this capability, with many more in the pipe.

1. The Sending Anchor makes a `SEP-12 GET /customer` request for the Sending and Receiving Clients if required.
1. The Sending Anchor makes a `SEP-12 PUT /customer` request for the Sending and Receiving Clients if required.
  1. The Receiving Anchor must validate the KYC data provided and reject the request with useful error messages if invalid.
1. The Sending Anchor may request a quote using the `SEP-38 GET /price` or `SEP-38 POST /quote` endpoints.
1. The Sending Anchor makes a `POST /transactions` request.

>   1. The Receiving Anchor must ensure the asset, amount, transaction fields, and customers IDs are valid. The Sending Anchor may provide a `quote_id` instead of asset and amount information if [SEP-38](sep-0038.md) is supported.

Author: I don't feel strongly either way. If we made asset & amount mandatory in all cases, the other optional parameters would be `quote_id` and `destination_asset`, which would be mutually exclusive. I don't like the idea of also requiring `destination_asset` in all requests since requiring a new field is a breaking change for existing implementations.

>   1. The Receiving Anchor must ensure the asset, amount, transaction fields, and customers IDs are valid. The Sending Anchor may provide a `quote_id` instead of asset and amount information if [SEP-38](sep-0038.md) is supported.

Reviewer 3: On second thought, requiring the user to repeat the fields that are already part of the quote goes against the [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself) principle. 

@leighmcculloch, @jacekn: if you both agree with that I'll approve this PR and update #973 to use the same strategy [here](https://github.com/stellar/stellar-protocol/pull/973/#discussion_r674033709).

>   1. The Receiving Anchor must ensure the asset, amount, transaction fields, and customers IDs are valid. The Sending Anchor may provide a `quote_id` instead of asset and amount information if [SEP-38](sep-0038.md) is supported.

Reviewer: It may be simpler, I think, to keep the API stable regardless if a quote is used, and to have the quote ID be an optional parameter. DRY isn't a truth, just a tool, and doesn't necessarily result in the simplest to understand APIs. In this specific case I do not feel strongly though and defer to you both.

Something that some APIs do to separate these changing input parameters from other parameters is to wrap them in a subfield and introduce a concept to wrap up those changing inputs. A good example of this is Stripe's source/payment intent interface. We could do something similar by saying the changing fields are the ExchangeParams, and for taht field you can either provide RawExchangeParams or QuoteExchangeParams.

>   1. The Receiving Anchor must ensure the asset, amount, transaction fields, and customers IDs are valid. The Sending Anchor may provide a `quote_id` instead of asset and amount information if [SEP-38](sep-0038.md) is supported.

Author: I really like this idea of a subfield, especially because the SEP-31 endpoint accepts `application/json` in a POST request. This pattern doesn't work well with SEP-6's `GET /*-exchange` endpoints though.

>   1. The Receiving Anchor must ensure the asset, amount, transaction fields, and customers IDs are valid. The Sending Anchor may provide a `quote_id` instead of asset and amount information if [SEP-38](sep-0038.md) is supported.

Reviewer: Even if we don't use a subfield, thinking about it more the subfield identifies a conceptual oddity with using mutual exclusion on destination_asset. For example:

<table>
<tr><th>non-quote</th><th>quote</th></tr>
<tr>
<td>

```json
{
  "asset": "...",
  "amount": "...",
  "subfield": {
    "destination_asset": "...",
  },
}
```

</td>
<td>

```json
{
  "asset": "...",
  "amount": "...",
  "subfield": {
    "quote_id": "...",
  },
}
```

</td>
</tr>
</table>

It's not intuitive to me that destination asset is inside the subfield when asset is outside. In the same way it's not intuitive that destination_asset is optional when asset isn't.

>   1. The Receiving Anchor must ensure the asset, amount, transaction fields, and customers IDs are valid. The Sending Anchor may provide a `quote_id` instead of asset and amount information if [SEP-38](sep-0038.md) is supported.

Author: Since we're not creating a `POST /transactions-exchange` endpoint and instead using the original `POST /transactions` endpoint we cannot require `destination_asset` on every request, otherwise that would be a breaking change for existing applications.

Personally I don't think having two mutually exclusive parameters is too confusing, especially if we have a subsection below the request parameters list explaining when to use each one.

>   1. The Receiving Anchor must ensure the asset, amount, transaction fields, and customers IDs are valid. The Sending Anchor may provide a `quote_id` instead of asset and amount information if [SEP-38](sep-0038.md) is supported.

Author: Updated language:

> The Receiving Anchor must ensure the asset, amount, transaction fields, and customers IDs are valid. The Sending Anchor may provide either, but not both, a `quote_id` or `destination_asset` attribute if [SEP-38](sep-0038.md) is supported.

>   1. The Receiving Anchor must ensure the asset, amount, transaction fields, and customers IDs are valid. The Sending Anchor may provide a `quote_id` instead of asset and amount information if [SEP-38](sep-0038.md) is supported.

Author: Actually, I just realized that `destination_asset` can be specified if `quote_id` is also specified. There is no need for mutual exclusion.

So the three combinations are:
- `asset_code`, `amount`
- `asset_code`, `amount`, `destination_asset`
- `asset_code`, `amount`, `destination_asset`, `quote_id`

And if `quote_id` is specified, the information must match that of the quote.

Thoughts?

}
```

The JSON object contains an entry for each Stellar asset that the Receiving Anchor supports receiving from the Sending Anchor. 

> If the Receiving Anchor supports [SEP-38](sep-0038.md), the Sending Client can check the [`SEP-38 GET /prices`](sep-0038.md#get-prices) endpoint to determine which off-chain assets the Receiving Anchor can deliver to recipients in exchange for an asset listed in the [`SEP-31 GET /info` response](#get-info). If the Receiving Anchor does not provide exchange rates for an asset, it is assumed that the Stellar asset can be exchanged 1-for-1 with the corresponding off-chain asset.

Reviewer 3: > If the Receiving Anchor does not provide exchange rates for an asset, it is assumed that the Stellar asset can be exchanged 1-for-1 with the corresponding off-chain asset.

Is that safe to assume? Are there anchors in the ecosystem operating that way?

Author: All existing anchors operate this way, since the Stellar asset they use is pegged to the fiat currency their users have.

  "sender_id": "d2bd1412-e2f6-4047-ad70-a1a2f133b25c",
  "receiver_id": "137938d4-43a7-4252-a452-842adcee474c",
  "fields": {
    "transaction": {
      "receiver_routing_number": "442928834",
      "receiver_account_number": "0029483242",
      "type": "SEPA"
    }
  }
}
```

##### Request Parameters

Name | Type | Description
-----|-----|------
`amount` | number | (optional) Amount of the Stellar asset sent to the Receiving Anchor.
`asset_code` | string | (optional) Code of the asset the Sending Anchor intends to send. This must match one of the entries listed in the receiving anchor's `GET /info` endpoint.
`asset_issuer` | string | (optional) The issuer of the Stellar asset the Sending Anchor intends to send. If not specified, the asset sent must be issued by the Receiving Anchor.
`destination_asset` | string | (optional) The off-chain asset the Receiving Anchor will deliver to the Receiving Client. The value must match one of the `asset` values included in a [`SEP-38 GET /prices?sell_asset=stellar:<asset_code>:<asset_issuer>`](sep-0038.md#get-prices) response using [SEP-38 Asset Identification Format](sep-0038.md#asset-identification-format).

> `quote_id` | string | The `id` returned from a `SEP-38 POST /quote` response. If this attribute is specified, none of the attributes listed above should also be included in the request. The Stellar transaction used to send the asset _must_ have a `created_at` timestamp that is earlier than the quote's `expires_at` attribute.

Reviewer 3: 🤔  Should we be including this `created_at` attribute you're referring to? Or should it actually be `started_at`?

```suggestion
`quote_id` | string | The `id` returned from a `SEP-38 POST /quote` response. If this attribute is specified, none of the attributes listed above should also be included in the request. The Stellar transaction used to send the asset _must_ have a `started_at` timestamp that is earlier than the quote's `expires_at` attribute.
```

Author: I'm referring to the `created_at` attribute returned from Horizon's `GET /transactions/:id` endpoint. This value is the close time of the ledger the transaction was included in. This gives us a source of trust to use so that there is never a dispute about whether the payment was made prior to the quote's expiration.

`status` | string | The status of the transaction. Values are outlined below.
`status_eta` | number | (optional) The estimated number of seconds until a status change is expected.
`amount_in` | string | (optional) The amount of the Stellar asset received or to be received by the Receiving Anchor. Excludes any fees charged after Receiving Anchor receives the funds.
`amount_in_asset` | string | (optional) The asset received or to be received by the Receiving Anchor. Must be present if `quote_id` or `destination_asset` was included in the `POST /transactions` request. The value must be in [SEP-38 Asset Identification Format](sep-0038.md#asset-identification-format).
`amount_out` | string | (optional) The amount sent by the Receiving Anchor to Receiving Client.
`amount_out_asset` | string | (optional) The asset delivered to the Receiving Client. Must be present if `quote_id` or `destination_asset` was included in the `POST /transactions` request. The value must be in [SEP-38 Asset Identification Format](sep-0038.md#asset-identification-format).
`amount_fee` | string | (optional) The amount of fee charged by the Receiving Anchor.

> `amount_fee_asset` | string | (optional) The asset in which fees are calculated in. Must be present if `quote_id` or `destination_asset` was included in the `POST /transactions` request. The value must be in [SEP-38 Asset Identification Format](sep-0038.md#asset-identification-format).

Reviewer 3: I believe the idea now is all conversion fees are included in the quote returned from SEP-38 /price and /quote endpoints, right? In that case, we don't need the attribute `amount_fee_asset` here.

Author: I think it's worth keeping as an optional attribute in cash a fixed fee is also charged on top of the margin included in the exchange rate. If only the margin on the exchange rate is used, then `amount_fee` will be 0 anyway.
